FCMExternalProberTarget/app/src/main/java/com/google/firebase/messaging/testing/fcmexternalprobertarget/FCMReceiveService.java

```diff
@@ -0,0 +1,109 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.messaging.testing.fcmexternalprobertarget;
+
+import android.content.Context;
+import android.content.Intent;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.localbroadcastmanager.content.LocalBroadcastManager;
+
+import com.google.firebase.messaging.FirebaseMessagingService;
+import com.google.firebase.messaging.RemoteMessage;
+
+import java.time.Clock;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+
+/**
+ * Service that handles interaction with Firebase Cloud Messaging, and logs the results of the
+ * messages both to the UI and files.
+ */
+public class FCMReceiveService extends FirebaseMessagingService {
+
+    private Context context;
+    private boolean viewLogging;
+    private Clock logTimer;
+
+    /**
+     * Create an instance for testing
+     * @param context Mocked Context object
+     * @param viewLogging Whether logs should be written to UI
+     */
+    public FCMReceiveService (Context context, boolean viewLogging, Clock logTimer) {
+        this.context = context;
+        this.viewLogging = viewLogging;
+        this.logTimer = logTimer;
+    }
+
+    public FCMReceiveService () {
+        this.context = this;
+        viewLogging = false;
+        logTimer = Clock.systemUTC();
+    }
+
+    @Override
+    public void onNewToken(@NonNull String token) {
+        // Store token in a file on external storage so it can be accessed by the probe
+        File tokenFile = makeExternalFile("","token.txt");
+        writeToFile(tokenFile, token);
+        logToUI("Info", "Token generated");
+    }
+
+    @Override
+    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
+        long receivedTime = logTimer.instant().getEpochSecond();
+        String sendTime = remoteMessage.getData().get("sendTime");
+        File logFile = makeExternalFile("logs", sendTime + ".txt");
+        writeToFile(logFile, Long.toString(receivedTime));
+        logToUI("Info","Message Received: " + sendTime + ".txt");
+    }
+
+
+    private void logToUI(String tag, String logText) {
+        Log.d(tag, logText);
+        if (viewLogging) {
+            return;
+        }
+        Intent intent = new Intent("updateUI");
+        intent.putExtra("logText", tag + ": " + logText);
+        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
+    }
+
+    private File makeExternalFile(String path, String fileName) {
+        File newPath = new File(context.getExternalFilesDir(null), path);
+        if (!newPath.exists()) {
+            if (!newPath.mkdirs()) {
+                logToUI("Error", "Unable to create directory " + path);
+            }
+            logToUI("Info", "Directory " + path + " created");
```

Comment:
This is a silent failure, since it always returns a new file whether or not it works.  

Since writeToFile also silently fails this is probably fine in this case, and other code isn't going to be built directly on top of this.  But generally this is a pattern to avoid.

Reply:
addressed in f88a918